api/v1/dois/PKPDoiHandler.php

```diff
@@ -269,28 +278,59 @@ public function edit(SlimRequest $slimRequest, APIResponse $response, array $arg
             return $response->withStatus(403)->withJsonError('api.dois.403.editItemOutOfContext');
         }
 
-        $params = $this->convertStringsToSchema(\PKP\services\PKPSchemaService::SCHEMA_DOI, $slimRequest->getParsedBody());
-        $params['id'] = $doi->getId();
+        $params = $this->convertStringsToSchema(PKPSchemaService::SCHEMA_DOI, $slimRequest->getParsedBody());
 
         $errors = Repo::doi()->validate($doi, $params);
         if (!empty($errors)) {
             return $response->withStatus(400)->withJson($errors);
         }
 
-        Repo::doi()->edit($doi, $params);
+        $pubObjectType = $slimRequest->getParsedBodyParam('pubObjectType');
+        $pubObjectId = $slimRequest->getParsedBodyParam('pubObjectId');
 
-        $doi = Repo::doi()->get($doi->getId());
+        // Default behaviour, only edits DOI
+        if (empty($pubObjectType) && empty($pubObjectId)) {
+            Repo::doi()->edit($doi, $params);
+            $doi = Repo::doi()->get($doi->getId());
 
-        return $response->withJson(Repo::doi()->getSchemaMap()->map($doi), 200);
+            return $response->withJson(Repo::doi()->getSchemaMap()->map($doi), 200);
+        }
+
+        $pubObjectRepo = $this->getPubObjectRepo($pubObjectType);
+        if (is_null($pubObjectRepo)) {
+            return $response->withStatus(403)->withJsonError('api.dois.403.pubTypeNotRecognized');
+        }
+
+        // Check pubObject for doiId
+        $pubObject = $pubObjectRepo->get($pubObjectId);
+        if ($pubObject?->getData('doiId') != $doi->getId()) {
+            return $response->withStatus(404)->withJsonError('api.dois.404.pubObjectNotFound');
+        }
+
+        // Copy DOI object data
+        $newDoi = clone $doi;
+        $newDoi->unsetData('id');
+        $newDoi->setAllData(array_merge($newDoi->getAllData(), ['doi' => $params['doi']]));
+        $newDoiId = Repo::doi()->add($newDoi);
+
+        // Update pubObject with new DOI and remove elsewhere if no longer in use
+        $pubObjectRepo->edit($pubObject, ['doiId' => $newDoiId]);
+        if (!Repo::doi()->isAssigned($doi->getId())) {
```

Comment:
We could theoretically provide pubObjectType to the isAssigned function -- the same DOI can only be assigned to an object from the same type -- or?

Reply:
That's a good idea. It would reduce the number of database queries for each check if it's only checking the one pub object type.